include/mata/nfa/nfa.hh

```diff
@@ -161,12 +161,31 @@ public:
      * @brief Get the useful states using a modified Tarjan's algorithm. A state
      * is useful if it is reachable from an initial state and can reach a final state.
      *
-     * @param state_at_first_useful_state Whether only the first found useful state is set to true. Useful when
-     *  checking whether the set of useful states is empty.
-     *
      * @return BoolVector Bool vector whose ith value is true iff the state i is useful.
      */
-    BoolVector get_useful_states(const bool stop_at_first_useful_state = false) const;
+    BoolVector get_useful_states() const;
+
+    /**
+     * @brief Structure for utilizing Tarjan's SCC discover.
+     * 
+     */
+    struct SCCCrawlExe {
+        // event handler for the first-time state discovery
+        std::function<bool(State)> state_discover;
+        // event handler for SCC discovery (together with the whole Tarjan stack)
+        std::function<bool(const std::vector<State>&, const std::vector<State>&)> scc_discover;
+        // event handler for state in SCC discovery
+        std::function<void(State)> scc_state_discover;
+        // event handler for visiting of the state successors
+        std::function<void(State,State)> succ_state_discover;
+    };
+
+    /**
+     * @brief Tarjan-based SCC crawler.
+     * 
+     * @param crawl Instance of the crawling class.
+     */
+    void scc_crawl(const SCCCrawlExe& crawl) const;
```

Comment:
It “crawls”, but what does the action of “crawling” mean in this context? I think an extended description would be useful here.

Reply:
Ok, I renamed the function (and some other things).